### ISSUE TITLE
In the DSR mode TCP reply may be sent from wrong IP when EP is reachable through multiple LB IPs

### ISSUE THREAD

**Issue by aleksey-novikov** [cilium/cilium#18632]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Lets make the following schema:
```
      +------+
      |Client| ----------+
      +------+           |
      /       \          |
+------+     +------+    |
| LB1  |     | LB2  |    |
+------+     +------+    |
      \       /          |
       +-----+           |
       | Pod | ----------+
       +-----+
```
When SYN packet sent from Client to LB1 in the BPF maps on the node that hosts Pod following entries are created:
- CT: (Client_IP,Client_Port)->(Pod_IP,Pod_Port)
- NAT: (Client_IP,Client_Port) -> (Pod_IP,Pod_Port) XLATE (LB1_IP, LB_Port)

When previous connection are closed (packet with FIN or RST flag set) by client or server side in the CT map entry the corresponding flag has been added - tx_closing or rx_closing. Next, on new SYN packet from the same Client_IP and Client_Port considered as CT_REOPENED and NAT map entry are replaced by new one. Good.

Imagine now that the FIN packet is lost and new SYN packet from the same Client_Port sent to LB2_IP. New connection has been balanced to the same EP Pod and, accordingly, to the same node. CT map entry does not has tx_closing or rx_closing state so the connection considered as CT_ESTABLISHED and the NAT map entry left unchanged. In this case reply packet will be SNATed to the LB1_IP which will cause the packet to be dropped on the client.

Thats all

### Cilium Version

1.10.7

### Kernel Version

4.18.0-305.3.1.el8.x86_64

### Kubernetes Version

1.20.5

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aleksey-novikov** [comment]
Packets sample: https://terabox.com/s/1nVpFgVDerDjE4aEB2iuqlA

**Comment by brb** [comment]
Thanks for the detailed issue. One potential fix is to extend this condition https://github.com/cilium/cilium/blob/v1.11/bpf/bpf_lxc.c#L1435 by adding the check "CT_ESTABLISHED && SYN is set".

**Comment by aleksey-novikov** [comment]
This code fragment is more suitable for changes [https://github.com/cilium/cilium/blob/v1.11/bpf/lib/conntrack.h#L255](https://github.com/cilium/cilium/blob/v1.11/bpf/lib/conntrack.h#L255)

**Comment by brb** [comment]
> This code fragment is more suitable for changes https://github.com/cilium/cilium/blob/v1.11/bpf/lib/conntrack.h#L255

If we modify this, then we need to make sure that we don't change a behavior of other code which relies on CT_REOPENED / CT_ESTABLISHED. Are you planning to work on it?

**Comment by aleksey-novikov** [comment]
> Are you planning to work on it?

No, I don't. We have already changed the deployment to avoid this issue

**Comment by brb** [comment]
From the commit message:
> This problem is likely to occur the right after users upgrade Cilium v1.11 to v1.12.

Setting the release blocker label accordingly.

**Comment by aleksey-novikov** [comment]
Similar issue #23126

**Comment by brb** [comment]
> If a CT_ESTABLISHED entry receives a TCP packet with DSR info (== SYN), we should now re-create the CT and SNAT state.

Isn't that we are just moving between which connection will be affected? Previously it was a new connection, while after #22978 - an old connection?

**Comment by julianwiedmann** [comment]
> > If a CT_ESTABLISHED entry receives a TCP packet with DSR info (== SYN), we should now re-create the CT and SNAT state.
> 
> @julianwiedmann Isn't that we are just moving between which connection will be affected? Previously it was a new connection, while after #22978 - an old connection? So, the underlying issue is still here.

I'm not 100% sure what the underlying issue is :sweat_smile: . But yes - if we get a fresh SYN, we essentially treat the old connection as stale ... that feels ok to me on first glance?

**Comment by brb** [comment]
By the underlying issue I meant that it's not possible to differentiate between the following concurrent flows:

* clientA:portA -> LB**1**:port -> backendB:portB
* clientA:portA -> LB**2**:port -> backendB:portB

However, this particular issue (https://github.com/cilium/cilium/issues/18632#issue-1115378394) should be resolved by your changes.

**Comment by julianwiedmann** [comment]
> By the underlying issue I meant that it's not possible to differentiate between the following concurrent flows:
> 
>     * clientA:portA -> LB**1**:port -> backendB:portB
> 
>     * clientA:portA -> LB**2**:port -> backendB:portB
> 
> 
:exploding_head: thanks, that drove the point home
